File: web/js/automaton.js
/**
 * Core data model for a finite automaton (DFA or NFA).
 *
 * Faithful JavaScript port of models/automaton.py from the desktop app.
 * A single class represents both DFA and NFA: an NFA is simply an
 * Automaton whose transition function maps to more than one target state
 * for some (state, symbol) pair, and/or uses epsilon transitions.
 */

const EPSILON = "ε";

/* ------------------------------------------------------------------ */
/* Small set helpers (Python set operators have no JS equivalent)      */
/* ------------------------------------------------------------------ */

function setUnion(a, b) {
  const out = new Set(a);
  for (const v of b) out.add(v);
  return out;
}

function setIntersection(a, b) {
  const out = new Set();
  for (const v of a) if (b.has(v)) out.add(v);
  return out;
}

function setDifference(a, b) {
  const out = new Set();
  for (const v of a) if (!b.has(v)) out.add(v);
  return out;
}

/** Deterministic ordering so labels and keys are reproducible. */
function sortedArray(iterable) {
  return Array.from(iterable).sort((x, y) => String(x).localeCompare(String(y), "en"));
}

/**
 * Separator for composite keys (state sets, transition rows). A control
 * character is used on purpose: state names are restricted to letters,
 * digits, "_" and "-", and generated names only add "{", "}" and ",", so
 * this byte can never appear inside one. That keeps "ab" and "a" + "b"
 * from ever producing the same key.
 */
const KEY_SEP = String.fromCharCode(1);

/** Canonical key for a set of states, standing in for Python frozenset. */
function setKey(iterable) {
  return sortedArray(iterable).join(KEY_SEP);
}

class Automaton {
  constructor(data) {
    const d = data || {};
    this.name = d.name !== undefined ? d.name : "Untitled";
    this.states = d.states ? d.states.slice() : [];
    this.alphabet = d.alphabet ? d.alphabet.slice() : [];
    this.startState = d.startState !== undefined ? d.startState : null;
    this.acceptStates = new Set(d.acceptStates || []);
    /** state -> symbol -> Set of target states */
    this.transitions = d.transitions || {};
    /** state -> [x, y] canvas coordinates */
    this.positions = d.positions || {};
  }

  /* ---------------------------------------------------------------- */
  /* State management                                                  */
  /* ---------------------------------------------------------------- */

  addState(name, position) {
    if (this.states.indexOf(name) !== -1) {
      throw new Error('State "' + name + '" already exists.');
    }
    this.states.push(name);
    this.transitions[name] = {};
    if (position) this.positions[name] = position;
    if (this.startState === null) this.startState = name;
  }

  removeState(name) {
    if (this.states.indexOf(name) === -1) return;
    this.states = this.states.filter((s) => s !== name);
    delete this.transitions[name];
    this.acceptStates.delete(name);
    delete this.positions[name];
    if (this.startState === name) {
      this.startState = this.states.length ? this.states[0] : null;
    }
    for (const src of Object.keys(this.transitions)) {
      for (const symbol of Object.keys(this.transitions[src])) {
        this.transitions[src][symbol].delete(name);
        if (this.transitions[src][symbol].size === 0) {
          delete this.transitions[src][symbol];
        }
      }
    }
  }

  renameState(oldName, newName) {
    if (oldName === newName) return;
    if (this.states.indexOf(newName) !== -1) {
      throw new Error('State "' + newName + '" already exists.');
    }
    const idx = this.states.indexOf(oldName);
    this.states[idx] = newName;
    this.transitions[newName] = this.transitions[oldName] || {};
    delete this.transitions[oldName];
    for (const src of Object.keys(this.transitions)) {
      for (const targets of Object.values(this.transitions[src])) {
        if (targets.has(oldName)) {
          targets.delete(oldName);
          targets.add(newName);
        }
      }
    }
    if (this.startState === oldName) this.startState = newName;
    if (this.acceptStates.has(oldName)) {
      this.acceptStates.delete(oldName);
      this.acceptStates.add(newName);
    }
    if (oldName in this.positions) {
      this.positions[newName] = this.positions[oldName];
      delete this.positions[oldName];
    }
  }

  setStartState(name) {
    if (this.states.indexOf(name) === -1) {
      throw new Error('State "' + name + '" does not exist.');
    }
    this.startState = name;
  }

  toggleAcceptState(name) {
    if (this.states.indexOf(name) === -1) {
      throw new Error('State "' + name + '" does not exist.');
    }
    if (this.acceptStates.has(name)) this.acceptStates.delete(name);
    else this.acceptStates.add(name);
  }

  /* ---------------------------------------------------------------- */
  /* Alphabet management                                               */
  /* ---------------------------------------------------------------- */

  addSymbol(symbol) {
    if (this.alphabet.indexOf(symbol) !== -1) {
      throw new Error('Symbol "' + symbol + '" already exists.');
    }
    this.alphabet.push(symbol);
  }

  removeSymbol(symbol) {
    if (this.alphabet.indexOf(symbol) === -1) return;
    this.alphabet = this.alphabet.filter((s) => s !== symbol);
    for (const src of Object.keys(this.transitions)) {
      delete this.transitions[src][symbol];
    }
  }

  /* ---------------------------------------------------------------- */
  /* Transition management                                             */
  /* ---------------------------------------------------------------- */

  addTransition(source, symbol, target) {
    if (this.states.indexOf(source) === -1) {
      throw new Error('Source state "' + source + '" does not exist.');
    }
    if (this.states.indexOf(target) === -1) {
      throw new Error('Target state "' + target + '" does not exist.');
    }
    if (symbol !== EPSILON && this.alphabet.indexOf(symbol) === -1) {
      throw new Error('Symbol "' + symbol + '" is not in the alphabet.');
    }
    if (!this.transitions[source]) this.transitions[source] = {};
    if (!this.transitions[source][symbol]) {
      this.transitions[source][symbol] = new Set();
    }
    this.transitions[source][symbol].add(target);
  }

  removeTransition(source, symbol, target) {
    const bySymbol = this.transitions[source];
    if (bySymbol && bySymbol[symbol]) {
      bySymbol[symbol].delete(target);
      if (bySymbol[symbol].size === 0) delete bySymbol[symbol];
    }
  }

  /** Targets reachable from source consuming exactly this symbol. */
  targets(source, symbol) {
    const bySymbol = this.transitions[source];
    if (!bySymbol || !bySymbol[symbol]) return new Set();
    return new Set(bySymbol[symbol]);
  }

  /** Flat [source, symbol, target] rows, ordered by the states list. */
  allTransitionRows() {
    const rows = [];
    for (const src of this.states) {
      const bySymbol = this.transitions[src] || {};
      for (const symbol of Object.keys(bySymbol)) {
        for (const dest of sortedArray(bySymbol[symbol])) {
          rows.push([src, symbol, dest]);
        }
      }
    }
    return rows;
  }

  /* ---------------------------------------------------------------- */
  /* Analysis helpers                                                  */
  /* ---------------------------------------------------------------- */

  hasEpsilonTransitions() {
    return this.states.some((s) => Boolean((this.transitions[s] || {})[EPSILON]));
  }

  isDeterministic() {
    if (this.hasEpsilonTransitions()) return false;
    for (const src of this.states) {
      const bySymbol = this.transitions[src] || {};
      for (const dests of Object.values(bySymbol)) {
        if (dests.size > 1) return false;
      }
    }
    return true;
  }

  isComplete() {
    if (!this.alphabet.length) return false;
    for (const state of this.states) {
      for (const symbol of this.alphabet) {
        if (this.targets(state, symbol).size === 0) return false;
      }
    }
    return true;
  }

  getReachableStates() {
    if (this.startState === null) return new Set();
    const seen = new Set([this.startState]);
    const stack = [this.startState];
    while (stack.length) {
      const state = stack.pop();
      const bySymbol = this.transitions[state] || {};
      for (const dests of Object.values(bySymbol)) {
        for (const d of dests) {
          if (!seen.has(d)) {
            seen.add(d);
            stack.push(d);
          }
        }
      }
    }
    return seen;
  }

  getUnreachableStates() {
    return setDifference(new Set(this.states), this.getReachableStates());
  }

  /** States from which no accept state can ever be reached. */
  getDeadStates() {
    const alive = new Set(this.acceptStates);
    const reverse = {};
    for (const s of this.states) reverse[s] = new Set();
    for (const src of this.states) {
      const bySymbol = this.transitions[src] || {};
      for (const dests of Object.values(bySymbol)) {
        for (const d of dests) {
          if (!reverse[d]) reverse[d] = new Set();
          reverse[d].add(src);
        }
      }
    }
    const stack = Array.from(alive);
    while (stack.length) {
      const state = stack.pop();
      for (const pred of reverse[state] || []) {
        if (!alive.has(pred)) {
          alive.add(pred);
          stack.push(pred);
        }
      }
    }
    return setDifference(new Set(this.states), alive);
  }

  /** True when every state is reachable from the start state. */
  isConnected() {
    return this.getUnreachableStates().size === 0;
  }

  validate() {
    const errors = [];
    if (!this.states.length) errors.push("The automaton has no states.");
    if (this.startState === null) {
      errors.push("No start state has been defined.");
    } else if (this.states.indexOf(this.startState) === -1) {
      errors.push("The start state is not in the list of states.");
    }
    if (!this.alphabet.length) errors.push("The alphabet is empty.");
    for (const a of sortedArray(this.acceptStates)) {
      if (this.states.indexOf(a) === -1) {
        errors.push('Accept state "' + a + '" does not exist.');
      }
    }
    for (const row of this.allTransitionRows()) {
      const src = row[0];
      const sym = row[1];
      if (sym !== EPSILON && this.alphabet.indexOf(sym) === -1) {
        errors.push("Transition (" + src + ", " + sym + ") uses a symbol outside the alphabet.");
      }
    }
    return errors;
  }

  /* ---------------------------------------------------------------- */
  /* Serialization - the .afjson format shared with the desktop app    */
  /* ---------------------------------------------------------------- */

  toJSON() {
    const transitions = {};
    for (const src of Object.keys(this.transitions)) {
      transitions[src] = {};
      for (const sym of Object.keys(this.transitions[src])) {
        transitions[src][sym] = sortedArray(this.transitions[src][sym]);
      }
    }
    const positions = {};
    for (const entry of Object.entries(this.positions)) {
      positions[entry[0]] = [entry[1][0], entry[1][1]];
    }
    return {
      name: this.name,
      states: this.states.slice(),
      alphabet: this.alphabet.slice(),
      start_state: this.startState,
      accept_states: sortedArray(this.acceptStates),
      transitions: transitions,
      positions: positions,
    };
  }

  static fromJSON(data) {
    const automaton = new Automaton({
      name: data.name !== undefined ? data.name : "Untitled",
      states: data.states || [],
      alphabet: data.alphabet || [],
      startState: data.start_state !== undefined ? data.start_state : null,
      acceptStates: data.accept_states || [],
    });
    const transitions = {};
    for (const entry of Object.entries(data.transitions || {})) {
      const src = entry[0];
      transitions[src] = {};
      for (const sub of Object.entries(entry[1])) {
        transitions[src][sub[0]] = new Set(sub[1]);
      }
    }
    automaton.transitions = transitions;
    for (const s of automaton.states) {
      if (!automaton.transitions[s]) automaton.transitions[s] = {};
    }
    const positions = {};
    for (const entry of Object.entries(data.positions || {})) {
      positions[entry[0]] = [entry[1][0], entry[1][1]];
    }
    automaton.positions = positions;
    return automaton;
  }

  clone() {
    return Automaton.fromJSON(this.toJSON());
  }
}

File: models/automaton.py
"""Core data model for a finite automaton (DFA or NFA).

A single class represents both DFA and NFA: an NFA is simply an
Automaton whose transition function maps to more than one target state
for some (state, symbol) pair, and/or uses epsilon transitions.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Set, Tuple

EPSILON = "ε"


@dataclass
class Automaton:
    """Generic finite automaton (DFA or NFA).

    transitions: state -> symbol -> set of target states.
    positions: state -> (x, y) canvas coordinates, used by the graph view.
    """

    name: str = "Sin título"
    states: List[str] = field(default_factory=list)
    alphabet: List[str] = field(default_factory=list)
    start_state: Optional[str] = None
    accept_states: Set[str] = field(default_factory=set)
    transitions: Dict[str, Dict[str, Set[str]]] = field(default_factory=dict)
    positions: Dict[str, Tuple[float, float]] = field(default_factory=dict)

    # ------------------------------------------------------------------
    # State management
    # ------------------------------------------------------------------
    def add_state(self, name: str, position: Optional[Tuple[float, float]] = None) -> None:
        if name in self.states:
            raise ValueError(f"El estado '{name}' ya existe.")
        self.states.append(name)
        self.transitions[name] = {}
        if position is not None:
            self.positions[name] = position
        if self.start_state is None:
            self.start_state = name

    def remove_state(self, name: str) -> None:
        if name not in self.states:
            return
        self.states.remove(name)
        self.transitions.pop(name, None)
        self.accept_states.discard(name)
        self.positions.pop(name, None)
        if self.start_state == name:
            self.start_state = self.states[0] if self.states else None
        for src in self.transitions:
            for symbol in list(self.transitions[src].keys()):
                self.transitions[src][symbol].discard(name)
                if not self.transitions[src][symbol]:
                    del self.transitions[src][symbol]

    def rename_state(self, old: str, new: str) -> None:
        if old == new:
            return
        if new in self.states:
            raise ValueError(f"El estado '{new}' ya existe.")
        idx = self.states.index(old)
        self.states[idx] = new
        self.transitions[new] = self.transitions.pop(old)
        for src in self.transitions:
            for symbol, targets in self.transitions[src].items():
                if old in targets:
                    targets.discard(old)
                    targets.add(new)
        if self.start_state == old:
            self.start_state = new
        if old in self.accept_states:
            self.accept_states.discard(old)
            self.accept_states.add(new)
        if old in self.positions:
            self.positions[new] = self.positions.pop(old)

    def set_start_state(self, name: str) -> None:
        if name not in self.states:
            raise ValueError(f"El estado '{name}' no existe.")
        self.start_state = name

    def toggle_accept_state(self, name: str) -> None:
        if name not in self.states:
            raise ValueError(f"El estado '{name}' no existe.")
        if name in self.accept_states:
            self.accept_states.discard(name)
        else:
            self.accept_states.add(name)

    # ------------------------------------------------------------------
    # Alphabet management
    # ------------------------------------------------------------------
    def add_symbol(self, symbol: str) -> None:
        if symbol in self.alphabet:
            raise ValueError(f"El símbolo '{symbol}' ya existe.")
        self.alphabet.append(symbol)

    def remove_symbol(self, symbol: str) -> None:
        if symbol not in self.alphabet:
            return
        self.alphabet.remove(symbol)
        for src in self.transitions:
            self.transitions[src].pop(symbol, None)

    # ------------------------------------------------------------------
    # Transition management
    # ------------------------------------------------------------------
    def add_transition(self, source: str, symbol: str, target: str) -> None:
        if source not in self.states:
            raise ValueError(f"El estado origen '{source}' no existe.")
        if target not in self.states:
            raise ValueError(f"El estado destino '{target}' no existe.")
        if symbol != EPSILON and symbol not in self.alphabet:
            raise ValueError(f"El símbolo '{symbol}' no pertenece al alfabeto.")
        self.transitions.setdefault(source, {}).setdefault(symbol, set()).add(target)

    def remove_transition(self, source: str, symbol: str, target: str) -> None:
        if source in self.transitions and symbol in self.transitions[source]:
            self.transitions[source][symbol].discard(target)
            if not self.transitions[source][symbol]:
                del self.transitions[source][symbol]

    def targets(self, source: str, symbol: str) -> Set[str]:
        return set(self.transitions.get(source, {}).get(symbol, set()))

    def all_transition_rows(self) -> List[Tuple[str, str, str]]:
        rows = []
        for src in self.states:
            for symbol, dests in self.transitions.get(src, {}).items():
                for dest in dests:
                    rows.append((src, symbol, dest))
        return rows

    # ------------------------------------------------------------------
    # Analysis helpers
    # ------------------------------------------------------------------
    def has_epsilon_transitions(self) -> bool:
        return any(EPSILON in self.transitions.get(s, {}) for s in self.states)

    def is_deterministic(self) -> bool:
        if self.has_epsilon_transitions():
            return False
        for src in self.states:
            for symbol, dests in self.transitions.get(src, {}).items():
                if len(dests) > 1:
                    return False
        return True

    def is_complete(self) -> bool:
        if not self.alphabet:
            return False
        for state in self.states:
            for symbol in self.alphabet:
                if not self.targets(state, symbol):
                    return False
        return True

    def get_reachable_states(self) -> Set[str]:
        if self.start_state is None:
            return set()
        seen = {self.start_state}
        stack = [self.start_state]
        while stack:
            state = stack.pop()
            for symbol, dests in self.transitions.get(state, {}).items():
                for d in dests:
                    if d not in seen:
                        seen.add(d)
                        stack.append(d)
        return seen

    def get_unreachable_states(self) -> Set[str]:
        return set(self.states) - self.get_reachable_states()

    def get_dead_states(self) -> Set[str]:
        """States from which no accept state can ever be reached."""
        alive: Set[str] = set(self.accept_states)
        changed = True
        # Build reverse adjacency
        reverse: Dict[str, Set[str]] = {s: set() for s in self.states}
        for src in self.states:
            for symbol, dests in self.transitions.get(src, {}).items():
                for d in dests:
                    reverse.setdefault(d, set()).add(src)
        stack = list(alive)
        while stack:
            state = stack.pop()
            for pred in reverse.get(state, set()):
                if pred not in alive:
                    alive.add(pred)
                    stack.append(pred)
        return set(self.states) - alive

    def is_connected(self) -> bool:
        """All states reachable from the start state."""
        return len(self.get_unreachable_states()) == 0

    def validate(self) -> List[str]:
        errors: List[str] = []
        if not self.states:
            errors.append("El autómata no tiene estados.")
        if self.start_state is None:
            errors.append("No se ha definido un estado inicial.")
        elif self.start_state not in self.states:
            errors.append("El estado inicial no existe en la lista de estados.")
        if not self.alphabet:
            errors.append("El alfabeto está vacío.")
        for a in self.accept_states:
            if a not in self.states:
                errors.append(f"El estado de aceptación '{a}' no existe.")
        for src, sym, dst in self.all_transition_rows():
            if sym != EPSILON and sym not in self.alphabet:
                errors.append(f"Transición ({src}, {sym}) usa un símbolo fuera del alfabeto.")
        return errors

    # ------------------------------------------------------------------
    # Serialization
    # ------------------------------------------------------------------
    def to_dict(self) -> dict:
        return {
            "name": self.name,
            "states": list(self.states),
            "alphabet": list(self.alphabet),
            "start_state": self.start_state,
            "accept_states": sorted(self.accept_states),
            "transitions": {
                src: {sym: sorted(dests) for sym, dests in syms.items()}
                for src, syms in self.transitions.items()
            },
            "positions": {s: list(p) for s, p in self.positions.items()},
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Automaton":
        automaton = cls(
            name=data.get("name", "Sin título"),
            states=list(data.get("states", [])),
            alphabet=list(data.get("alphabet", [])),
            start_state=data.get("start_state"),
            accept_states=set(data.get("accept_states", [])),
        )
        automaton.transitions = {
            src: {sym: set(dests) for sym, dests in syms.items()}
            for src, syms in data.get("transitions", {}).items()
        }
        for s in automaton.states:
            automaton.transitions.setdefault(s, {})
        automaton.positions = {
            s: tuple(p) for s, p in data.get("positions", {}).items()
        }
        return automaton

    def clone(self) -> "Automaton":
        return Automaton.from_dict(self.to_dict())
